### PR TITLE
 Retryable SubmitForm686cJob and SubmitForm674Job (Pending Full Idempotence)

### PR DESCRIPTION
**Pull request by tfink419**
## Summary
- Allow `SubmitForm686cJob` and `SubmitForm674Job` sidekiq jobs to be retryable up to 14 times
  - Log error after each attempt but only send email and attempt to recover form on final attempt
- Partial idempotence (only if there are errors)
- Move error handling to only send email and save form after retries are exhausted
  - Additionally move user object logic to top of jobs so that error handling can always have access to user object

## Related issue(s)
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/65052
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/56290
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/65151

## Testing done
- Create a claim via /view-change-dependents/add-remove-form-21-686c/
- Check in /sidekiq admin console that claim will retry if there is an error (locally it will fail always)

## Screenshots
Screenshot 2023-09-19 at 16 46 40

## What areas of the site does it impact?
None are functionally impacted, this should only assist in background jobs being handled better after errors

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [x]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [x]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [x]  I added a screenshot of the developed feature

### CODE REVIEW

File: app/workers/bgs/submit_form674_job.rb [hunk 1]
```diff
@@ -9,37 +9,48 @@ class Invalid674Claim < StandardError; end
...
     include SentryLogging
 
-    sidekiq_options retry: false
+    attr_reader :claim, :user
+
+    sidekiq_options retry: 14
+
+    sidekiq_retries_exhausted do
+      salvage_save_in_progress_form(FORM_ID, @user_uuid, @in_progress_copy)
+      DependentsApplicationFailureMailer.build(user).deliver_now if user&.email.present?
```
Comment (tblackwe) [inline app/workers/bgs/submit_form674_job.rb:18]: I think this is tickets out of sync, shouldnt we be leveraging the backup submission process here?

File: lib/bgs/form674.rb [hunk 2]
```diff
@@ -39,22 +42,33 @@ def submit(payload)
...
+        bgs_service.update_proc(proc_id, proc_state: 'MANUAL_VAGOV')
+      rescue
+        Rails.logger.warning('BGS::Form674.submit failed after creating benefit claim in BGS',
+                             {
+                               user_uuid: user.uuid,
+                               saved_claim_id: saved_claim.id,
+                               icn: user.icn,
+                               error: e.message
+                             })
+        log_message_to_sentry(e, :warning, {}, { team: 'vfs-ebenefits' })
```
Comment (tblackwe) [inline lib/bgs/form674.rb:70]: Rails.logger.warning should be sufficient to get the data to sentry.

File: lib/bgs/form686c.rb [hunk 3]
```diff
@@ -49,19 +52,31 @@ def submit(payload)
...
+        bgs_service.update_proc(proc_id, proc_state: @proc_state)
+      rescue => e
+        Rails.logger.warning('BGS::Form686c.submit failed after creating benefit claim in BGS',
+                             {
+                               user_uuid: user.uuid,
+                               saved_claim_id: saved_claim.id,
+                               icn: user.icn,
+                               error: e.message
+                             })
+        log_message_to_sentry(e, :warning, {}, { team: 'vfs-ebenefits' })
```
Comment (tblackwe) [inline lib/bgs/form686c.rb:78]: Rails.logger.warning should be sufficient to get the data to sentry.

**Review by tblackwe — CHANGES_REQUESTED** [review]

File: app/sidekiq/bgs/submit_form674_job.rb [hunk 4]
```diff
@@ -9,45 +9,52 @@ class Invalid674Claim < StandardError; end
...
+      raise
     end
-    # rubocop:enable Metrics/MethodLength
 
     private
 
-    def valid_claim_data(saved_claim_id, vet_info)
-      claim = SavedClaim::DependencyClaim.find(saved_claim_id)
+    def valid_claim_data
+      @claim = SavedClaim::DependencyClaim.find(saved_claim_id)
```
Comment (evansmith) [inline app/sidekiq/bgs/submit_form674_job.rb:57]: with claim in the attr_reader, is getting @claim still necessary?

File: app/sidekiq/bgs/submit_form686c_encrypted_job.rb [hunk 5]
```diff
@@ -9,43 +9,54 @@ class Invalid686cClaim < StandardError; end
...
+
+      raise
     end
 
     private
 
-    def valid_claim_data(saved_claim_id, vet_info)
-      claim = SavedClaim::DependencyClaim.find(saved_claim_id)
+    def valid_claim_data
+      @claim = SavedClaim::DependencyClaim.find(saved_claim_id)
```
Comment (evansmith) [inline app/sidekiq/bgs/submit_form686c_encrypted_job.rb:59]: same claim question as above

File: app/sidekiq/bgs/submit_form686c_job.rb [hunk 6]
```diff
@@ -9,47 +9,55 @@ class Invalid686cClaim < StandardError; end
...
+      raise
     end
-    # rubocop:enable Metrics/MethodLength
 
     private
 
-    def valid_claim_data(saved_claim_id, vet_info)
-      claim = SavedClaim::DependencyClaim.find(saved_claim_id)
+    def valid_claim_data
+      @claim = SavedClaim::DependencyClaim.find(saved_claim_id)
```
Comment (evansmith) [inline app/sidekiq/bgs/submit_form686c_job.rb:60]: same claim question as above

File: app/sidekiq/bgs/submit_form674_encrypted_job.rb [hunk 7]
```diff
@@ -9,47 +9,55 @@ class Invalid674Claim < StandardError; end
...
+      raise
     end
-    # rubocop:enable Metrics/MethodLength
 
     private
 
-    def valid_claim_data(saved_claim_id, vet_info)
-      claim = SavedClaim::DependencyClaim.find(saved_claim_id)
+    def valid_claim_data
+      @claim = SavedClaim::DependencyClaim.find(saved_claim_id)
```
Comment (evansmith) [inline app/sidekiq/bgs/submit_form674_encrypted_job.rb:60]: with claim in the attr_reader, is getting @claim still necessary?

**Review by evansmith — COMMENTED** [review]
looks good and the tests pass. just had a quick question about a var.

File: app/workers/bgs/submit_form674_job.rb (hunk 1 above)
Comment (tfink419) [reply app/workers/bgs/submit_form674_job.rb]: Yes, it should be updated to reflect the backup submission now

File: app/sidekiq/bgs/submit_form674_job.rb (hunk 4 above)
Comment (tfink419) [reply app/sidekiq/bgs/submit_form674_job.rb:57]: creating an instance variable is necessary for this line
https://github.com/department-of-veterans-affairs/vets-api/blob/827b92032ed9f5d093ac9cc9fe3b87b8f445ef13/app/sidekiq/bgs/submit_form674_encrypted_job.rb#L45

**Review by evansmith — DISMISSED** [review]
LGTM

**Review by tblackwe — APPROVED** [review]

**Review by ryan-mcneil — APPROVED** [review]
👌

**Pull request merged**